### ISSUE TITLE
How to make the most of compute cluster?

### ISSUE THREAD

**Issue by heyheys** [Azure/azureml-examples#1790]
Hi there,

I've got a question and couldn't find answer online. Appreciate if someone can answer it!

Question: To make to most of compute cluster with multiple nodes, what are the requirements on the framework/library used?

For example, I think if I process Pandas DataFrame in the source code, only one node is actually being used even if I set the computer cluster with multiple nodes. Is this true? If so, does this mean to make to most of the compute cluster, the framework/library we're using should provide multiple-core handling, and we need to set some parameters (eg. num_of_threads) in the source code such that it aligns with the number of nodes in the compute cluster? If so, can you name some common libraries that can make advantage of compute cluster? In particular, I'd like to know if Pandas, LightGBM, PyTorch can use multiple nodes in the compute cluster.

Many thanks!

**Comment by mikkeyboi** [comment]
A higher level example is that more nodes are spun up when `PythonScriptStep`s are executed in parallel. When multiple nodes are working together to complete multiple workflows, this means that each node can execute the code independently.

In a finer grained scenario, such as pandas operations, recruiting more nodes mean that their extractions or transformations are not dependent on one another. One way to achieve this is to make multiple segments of your data (could be as simple as using the sample dataset but indexing at different ranges) and populate the job.
`preprocess_1 = [a1, a2, a3, a4, a5, a6]`
`preprocess_2 = [b1, b2, b3, b4, b5, b6]`
`StepSequence(steps=preprocess_1, preprocess_2)` where a[n] or b[n] are `PythonScriptStep` or similar.
This would recruit 6 nodes to complete `preprocess_1` and `preprocess_2`.

Or you could do another approach you were probably mentioning and use Dask or Modin. PyTorch you have to load your models into the correct GPUs if you choose a mutli-GPU node.
If you're mostly working with tables, you can try Delta tables and run your workloads on Spark clusters (Synapse, Databricks, etc.). These would scale naturally when drivers ask the cluster manager to recruit executors for each job.

**Comment by heyheys** [comment]
Really appreciate you answer!
